Add `tools` property and entrypoint for workflows

bastonero opened the pull request:
Fixes #6865

agoscinski commented:
@bastonero sorry I did not give you heads up. I wanted to merge it and was about to finalize it but then there was an error in the REST API which I was not able to fix in the time

agoscinski commented:
@eimrek can you have a look? It is a test error in the REST API

mbercx commented:
I'm very interested in having this feature. Adding tools to my work chains would be sexy AF. Any chance we can get this merged soon? @eimrek can you look into the blocking issue during the coding week?

bastonero commented:
Hey @agoscinski @eimrek @mbercx, do you think we could merge this one anytime soon? It would be really useful to have.

bastonero committed:
✨ Add workflow tool entry points (#6884)

Add a `tools` property to `WorkChainNode` and load workflow helpers
from `aiida.tools.workflows` entry points.

Register the new workflow tools entry point group, add the base
`WorkflowTools` implementation, and cover the workflow and calculation
fallback paths with tests.

Co-authored-by: Alexander Goscinski <alex.go@posteo.de>
Co-authored-by: Kristjan Eimre <eimrek@users.noreply.github.com>

danielhollas reviewed (commented):
Sorry I won't have time to look at this anytime soon so leaving review to others.

mbercx commented:
I'm keen to have this feature, so happy to review later this week (I think Thu I will have time). Will also field-test it on a workflow I wanted it for. Just have to remember which one that was. 😅 

In any case: @bastonero I don't think I see any changes to the documentation. Should we add this somewhere, or update the docs that mention this for `CalcJob`s?

agoscinski commented:
The flaky test failed `tests/cmdline/commands/test_group.py` that does wrong pattern matching because it depends on the time it takes.
```
tests/cmdline/commands/test_group.py:368: in test_show_limit
    assert not (str(nodes[0].pk) in result.output and str(nodes[1].pk) in result.output)
E   AssertionError: assert not ('1' in '-----------------  ----------------\nGroup label        test_group_limit\nGroup type_string  core\nGroup description ...n-----------------  ----------------\n# Nodes:\n  PK  Type    Created\n----  ------  ---------\n   1  Data    2s ago\n' and '2' in '-----------------  ----------------\nGroup label        test_group_limit\nGroup type_string  core\nGroup description ...n-----------------  ----------------\n# Nodes:\n  PK  Type    Created\n----  ------  ---------\n   1  Data    2s ago\n')
E    +  where '1' = str(1)
E    +    where 1 = <Data: uuid: 04be5ac9-6213-452f-869f-9a035df08693 (pk: 1)>.pk
E    +  and   '-----------------  ----------------\nGroup label        test_group_limit\nGroup type_string  core\nGroup description ...n-----------------  ----------------\n# Nodes:\n  PK  Type    Created\n----  ------  ---------\n   1  Data    2s ago\n' = CliResult(stderr_bytes=b'', stdout_bytes=b'-----------------  ----------------\nGroup label        test_group_limit\nG...   Created\n----  ------  ---------\n   1  Data    2s ago\n', exc_info=(None, None, None), exception=None, exit_code=0).output
E    +  and   '2' = str(2)
E    +    where 2 = <Data: uuid: 787eee60-4885-45ad-a386-b4330f085c19 (pk: 2)>.pk
E    +  and   '-----------------  ----------------\nGroup label        test_group_limit\nGroup type_string  core\nGroup description ...n-----------------  ----------------\n# Nodes:\n  PK  Type    Created\n----  ------  ---------\n   1  Data    2s ago\n' = CliResult(stderr_bytes=b'', stdout_bytes=b'-----------------  ----------------\nGroup label        test_group_limit\nG...   Created\n----  ------  ---------\n   1  Data    2s ago\n', exc_info=(None, None, None), exception=None, exit_code=0).output
```
Currently it is fixed it PR #7325 in commit fe23aede5b but I need to make another PR for it. I will ignore therefore ignore the error of CI ci-code / test minimum reqs (3.10, sqlite)
 
---
The other problem is `tests/engine/test_launch.py::test_await_processes`in  need to check it in `tests (3.11)` CI. The error is however flaky and most likely not related to the changes in this PR. I noted it down the error and will try fix it.
